### PR TITLE
Add Group Add and Remove acks

### PR DESCRIPTION
**Pull request by BrennanConroy**
We'll probably want to factor this out of Redis and make it generic to HubLifetimeManager after alpha

https://github.com/aspnet/SignalR/issues/584

### CODE REVIEW

File: src/Microsoft.AspNetCore.SignalR.Redis/RedisHubLifetimeManager.cs [hunk 1]
```diff
@@ -108,6 +111,43 @@ public class RedisHubLifetimeManager<THub> : HubLifetimeManager<THub>, IDisposab
...
+                    }
+                }
+                else if (groupMessage.Action == GroupAction.Ack)
+                {
+                    _ackHandler.TriggerAck(groupMessage.Id);
+                    return;
+                }
+
+                // Sending ack
+                await PublishAsync(channelName, new GroupMessage
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.SignalR.Redis/RedisHubLifetimeManager.cs]: Send to just the server that is waiting for the Ack

File: src/Microsoft.AspNetCore.SignalR.Redis/RedisHubLifetimeManager.cs [hunk 2]
```diff
@@ -241,11 +281,22 @@ public override Task OnDisconnectedAsync(HubConnectionContext connection)
...
+                // short circuit if connection is on this server
+                return;
+            }
+
+            await SendGroupActionAndWaitForAck(connectionId, groupName, GroupAction.Add);
+        }
+
+        private async Task<bool> AddGroupAsyncCore(string connectionId, string groupName)
+        {
             var groupChannel = _channelNamePrefix + ".group." + groupName;
```
Comment (moozzyk) [inline src/Microsoft.AspNetCore.SignalR.Redis/RedisHubLifetimeManager.cs]: nit: this should be moved closer to the usage. We are doing this but may not use at all if the connection with the given id does not exist (same in RemoveGroupAsync)

File: src/Microsoft.AspNetCore.SignalR.Redis/RedisHubLifetimeManager.cs [hunk 3]
```diff
@@ -420,5 +506,91 @@ private class RedisFeature : IRedisFeature
...
+
+            public bool TriggerAck(int id)
+            {
+                if (_acks.TryRemove(id, out var ack))
+                {
+                    ack.Tcs.TrySetResult(null);
+                    return true;
+                }
+
+                return false;
```
[the remark this answers is not in the document]
Comment (BrennanConroy) [reply src/Microsoft.AspNetCore.SignalR.Redis/RedisHubLifetimeManager.cs]: Copied from old SignalR, look like the value was only used in tests.

File: src/Microsoft.AspNetCore.SignalR.Redis/Internal/AckHandler.cs [hunk 4]
```diff
@@ -0,0 +1,85 @@
+using System;
```
Comment (mikaelm12) [inline src/Microsoft.AspNetCore.SignalR.Redis/Internal/AckHandler.cs]: Header

File: src/Microsoft.AspNetCore.SignalR.Redis/Internal/AckHandler.cs [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        public DateTime Created { get; private set; }
+
+        public AckInfo()
+        {
+            Created = DateTime.UtcNow;
+            Tcs = new TaskCompletionSource<object>(TaskContinuationOptions.RunContinuationsAsynchronously);
+        }
+    }
+
+    internal enum GroupAction
```
Comment (mikaelm12) [inline src/Microsoft.AspNetCore.SignalR.Redis/Internal/AckHandler.cs]: Is this the best place for this?

File: src/Microsoft.AspNetCore.SignalR.Redis/RedisHubLifetimeManager.cs [hunk 6]
```diff
@@ -420,5 +517,21 @@ private class RedisFeature : IRedisFeature
             public HashSet<string> Subscriptions { get; } = new HashSet<string>();
             public HashSet<string> Groups { get; } = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
         }
+
+        private enum GroupAction
```
Comment (mikaelm12) [inline src/Microsoft.AspNetCore.SignalR.Redis/RedisHubLifetimeManager.cs]: This feels better 👍

File: src/Microsoft.AspNetCore.SignalR.Redis/Internal/AckHandler.cs [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+                    if (_acks.TryRemove(pair.Key, out var ack))
+                    {
+                        ack.Tcs.TrySetCanceled();
+                    }
+                }
+            }
+        }
+    }
+
+    internal class AckInfo
```
Comment (moozzyk) [inline src/Microsoft.AspNetCore.SignalR.Redis/Internal/AckHandler.cs]: Can this be private in AckHandler - does not seem to be used outside this class

File: src/Microsoft.AspNetCore.SignalR.Redis/Internal/AckHandler.cs [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+                return _acks.GetOrAdd(id, _ => new AckInfo()).Tcs.Task;
+            }
+        }
+
+        public bool TriggerAck(int id)
+        {
+            if (_acks.TryRemove(id, out var ack))
+            {
+                ack.Tcs.TrySetResult(null);
+                return true;
```
Comment (moozzyk) [inline src/Microsoft.AspNetCore.SignalR.Redis/Internal/AckHandler.cs]: I would just remove return value for now

File: src/Microsoft.AspNetCore.SignalR.Redis/Internal/AckHandler.cs [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+            _timer = new Timer(_ => CheckAcks(), state: null, dueTime: _ackInterval, period: _ackInterval);
+        }
+
+        public Task CreateAck(int id)
+        {
+            lock (_lock)
+            {
+                if (_disposed)
+                {
+                    return Task.CompletedTask;
```
Comment (moozzyk) [inline src/Microsoft.AspNetCore.SignalR.Redis/Internal/AckHandler.cs]: I think this causes issues - you will return a completed task and will still try to send a message over Redis (which is probably disposed as well)...

Comment (BrennanConroy) [reply src/Microsoft.AspNetCore.SignalR.Redis/Internal/AckHandler.cs]: Thats fine, you'll just get whatever error you normally would have gotten when trying to publish a message from a disposed redis instance. This doesn't change anything

Comment (moozzyk) [reply src/Microsoft.AspNetCore.SignalR.Redis/Internal/AckHandler.cs]: I was wondering if we should not have a `_disposed` flag in the RedisLifetimeManager and throw/not handle requests after it was disposed (not really sure how likely it is to happen)

Comment (BrennanConroy) [reply src/Microsoft.AspNetCore.SignalR.Redis/Internal/AckHandler.cs]: We might want to consider this for preview, but don't think it is related to this change.

**Comment by BrennanConroy** [comment]
Ping

File: src/Microsoft.AspNetCore.SignalR.Redis/Internal/AckHandler.cs (hunk 9 above)
Comment (moozzyk) [reply src/Microsoft.AspNetCore.SignalR.Redis/Internal/AckHandler.cs]: Sound good.

**Pull request merged**